src/buildout/wheel/__init__.py

```diff
@@ -1,56 +1,222 @@
-import distlib.wheel
-import humpty
+import glob
+import logging
 import os
+import os.path
+import shutil
+import sys
+import tempfile
 import pkg_resources
 import setuptools.package_index
-from six.moves.urllib import parse
 import wheel.install
+import wheel.decorator
 import pip.pep425tags
 import zc.buildout.easy_install
+import distutils.command.install
 
-orig_interpret_distro_name = setuptools.package_index.interpret_distro_name
+
+logger = logging.getLogger(__name__)
+
+NAMESPACE_STUB_PATH = os.path.join(os.path.dirname(__file__),
+                                   'namespace_stub.py')
+assert os.path.isfile(NAMESPACE_STUB_PATH)
+
+
+orig_distros_for_location = setuptools.package_index.distros_for_location
 original_wheel_to_egg = zc.buildout.easy_install.wheel_to_egg
+orig_Installer = zc.buildout.easy_install.Installer
+
+
+def _get_dest_dist_paths(dest):
+    p = os.path
+    eggs = glob.glob(p.join(dest, '*.egg'))
+    dists = [p.dirname(dist_info) for dist_info in
+             glob.glob(p.join(dest, '*', '*.dist-info'))]
+    # sort them like pkg_resources.find_on_path() would
+    return pkg_resources._by_version_descending(set(eggs + dists))
+
+
+class Environment(pkg_resources.Environment):
```

Comment:
> I'm worried about these cross-org subclasses.

There are two parts to the `Environment` subclass.

The first is that, as of today, Buildout passes the `eggs-directory` to `Environment` as a single path, instead of passing all eggs as separate paths.

I think this is abusing the `Environment` a little, since `eggs-directory` is never going to be directly in `sys.path`, but `Environment` treats it as it it would.

If we change buildout to feed all eggs individually to `Environment`, the `scan()` override wouldn't be necessary.

The second aspect is a bit trickier. Buildout treats specially all dists that have `EGG_DIST` precedence. It assumes they're things it can chose to use or ignore and fetch something more appropriate. Whereas any other `precedence` value is treated by Buildout as things that it can't change (i.e. they're either develop eggs, or come from the environment).

In order not to override the `Environment.add` method, the `Installer` would need to check each `dist.location` against the `eggs-directory` instead of checking the precedence.

That or we would need to go through all dists in `Environment` and mark those inside `eggs-directory` as `EGG_DIST` after each `scan`.

So, if we're willing to change buildout a bit, the `Environment` subclass would be unnecessary.